### DISCUSSION TITLE
Read consistency across keys

### DISCUSSION THREAD

**Discussion by plaisted** [microsoft/garnet#116] [Q&A] [answered]
I know this is not a relational database but are there any capabilities / strategies for getting consistent reads across keys other than locking? In this case I don't care if the values are updated during the read I just want to see a consistent snapshot of the values at a point in time (similar to a select query in a database, eg. postgres, with read committed isolation). The groups of keys to be read together consistently would be known on insert so could be organized in a way to support this (eg. key naming for hashing etc).

If not are there limitations / recommended upper bound on number of keys to be shared locked at same time?

**Comment by badrishc** [comment, marked as answer]
You can wrap the reads into a transaction, then you will get a consistent read across the keys. Performance should be good for many keys, as long are there isn't a lot of contention with writes.

https://microsoft.github.io/garnet/docs/extensions/transactions

**Comment by plaisted** [comment]
Any non-transactional alternatives, eg. if using the AOF / WAL are there any utilities to scan this as a background process to create a consistent non-write blocking view over the database?

**Reply by badrishc** [reply]
If you enable replication, the replica reads the AOF and serves a read-only copy of the store.
